### PR TITLE
Add new Rich Nav specific pipeline, add rich nav to test CI

### PR DESCRIPTION
**Pull request by jepetty**
I've altered two files:
1. Add a Rich Navigation specific build (which will appear as a new build pipeline), indexing on every push to master. This will allow you to use Rich Navigation in Visual Studio to enable cross-repository searches, serverless navigation at https://online.visualstudio.com/github/dotnet/msbuild, and integrated GitHub C# navigation once that is released (or, comment below if you want to be added to the preview now 😊).
1. Added the Rich Navigation task to your normal unit test CI. This has all the same benefits of ^, but also indexes your pull requests for a richer code review experience. This does come at the cost of increasing the time of your CI.

Based on your team's preference, only one of these changes is necessary.

## Rich Code Navigation

[Docs Link](https://devdiv.visualstudio.com/DevDiv/_wiki/wikis/DevDiv.wiki/1047/Rich-Code-Navigation)
With Rich Code Navigation, you can use editor-level navigation capabilities (like peek definition, find all references, and even diagnostics) on a pull request, without requiring a local checkout. This is available across all files and dependencies of your repository.

Rich Code Navigation is available for Visual Studio (for repos hosted on GitHub or Azure Repos) and Visual Studio Code (for repos hosted on GitHub).

Rich Code Navigation supports these capabilities:

- Hover
- Peek definition and Go to definition
- Peek references and Find all references
- Go to type definition
- Go to implementation
- Diagnostics

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/project-system/pull/6665)

### CODE REVIEW

**Comment by jepetty** [comment]
@marcpopMSFT FYI!

**Comment by drewnoakes** [comment]
> comment below if you want to be added to the preview now 😊

I'd be keen to try this out, for sure.

**Comment by drewnoakes** [comment]
> This does come at the cost of increasing the time of your CI.

Do we have a sense of how much longer we're talking? The legs on this build seem to have taken about the usual time to complete.

File: build/ci/unit-tests-template.yml [hunk 1]
```diff
@@ -7,6 +8,13 @@ jobs:
     - script: $(Build.SourcesDirectory)\build.cmd /build /test /ci /sign /diagnostic /clearnugetcache /no-deploy /no-integration /no-ibc /configuration ${{ parameters.configuration }}
       displayName: Build ProjectSystem.sln
 
+    - ${{ if eq(parameters.enableRichCodeNavigation, true) }}:
+      - task: RichCodeNavIndexer@0
+        displayName: RichCodeNav Upload
+        inputs:
+          languages: 'csharp'
```
Comment (dmonroym) [inline build/ci/unit-tests-template.yml:15]: We also have a lot of code on VB, is RichNav capable of that?

File: build/ci/unit-tests.yml [hunk 2]
```diff
@@ -29,6 +29,7 @@ jobs:
   parameters:
     name: Windows_Debug
     configuration: Debug
+    enableRichNavigation: true
```
Comment (dmonroym) [inline build/ci/unit-tests.yml:32]: Should this be enableRichCodeNavigation?

**Comment by dmonroym** [comment]
> > This does come at the cost of increasing the time of your CI.
> 
> Do we have a sense of how much longer we're talking? The legs on this build seem to have taken about the usual time to complete.
> 
> Looking through this PR's build logging, I see no indication of what extra steps were taken or how long they took.

@drewnoakes it doesn't seem to have ran. @jepetty it seems like you're re-building on the richcode nav task again though, this would mean almost doubling of our build time, is it possible to do it in one pass?

File: build/ci/unit-tests.yml (hunk 2 above)
Comment (jepetty) [reply build/ci/unit-tests.yml:32]: Yes, good catch! That's why the indexing step didn't run, my bad. I'll respond to the thread once it runs to let you know how much time _might_ be added to your main CI

File: build/ci/unit-tests-template.yml (hunk 1 above)
Comment (jepetty) [reply build/ci/unit-tests-template.yml:15]: Yes it is! With support from the Roslyn team, both C# and VB are ingested :)

Comment (jepetty) [reply build/ci/unit-tests-template.yml:15]: It's great to know your team utilizes VB. Once we get y'all set up with the clients, we'd love to hear more of your feedback on the VB experience :)

**Comment by jepetty** [comment]
It looks like the Rich Nav task would add just 5 minutes to your regular CI. @drewnoakes and @dmonroym let me know if you have any questions about the pros and cons of adding it to your regular CI vs creating an entirely new pipeline!

**Comment by tmeschter** [comment]
I'm in favor of having this turned on for PRs, at least to try out. It sounds like we can always step down to option 1 if we decide the extra time isn't worth it, right?

**Comment by tmeschter** [comment]
I'm in favor of having this turned on for PRs, at least to try out. It sounds like we can always step down to option 1 if we decide the extra time isn't worth it, right?

**Comment by jepetty** [comment]
@tmeschter yup, will be a fairly easy switch if you find it's taking too much time and not worth the benefit for PRs. If @drewnoakes or @dmonroym don't object, I'll remove the Rich Nav specific pipeline

**Comment by drewnoakes** [comment]
No objection from me, though @dmonroym spends more time with our infra so I'd hold out for his input.

**Comment by dmonroym** [comment]
@jepetty where did you get the 5 mins from? I believe it's only 1 min for that step. Overall I think it's also worth adding and it would be great if the team could get a demo of the advantages of it and how to use it - @jjmew perhaps this is something we can do for our team meeting.

I do think that we should keep it as a different pipeline though so it doesn't block insertions as we test it out and then once it's been properly baked we can move it back to how it currently is working and just keep it in the debug pipeline.

**Comment by jepetty** [comment]
@dmonroym sorry - I got it from [this build](https://dev.azure.com/dnceng/public/_build/results?buildId=847584&view=logs&j=908bce63-940e-59f8-7204-204b1f5ead0b&t=7abe0d25-4849-5f72-be2e-aed83ed4f6f10).

Sounds good though! I've removed the steps from your normal pipeline. I can't push this branch to dotnet/project-system, so I'll create a RichNav specific pipeline once this merges

File: build/ci/unit-tests-template.yml [hunk 3]
```diff
@@ -1,4 +1,5 @@
 # Template used by unit-tests.yml
+
```
Comment (dmonroym) [inline build/ci/unit-tests-template.yml:2]: nit: remove this file

File: build/ci/richnav.yml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    solution: 'build\proj\internal\RestoreIBCMerge.csproj'
+    feed: '8f470c7e-ac49-4afe-a6ee-cf784e438b93'
+
+- script: $(Build.SourcesDirectory)\build.cmd /build /ci /no-deploy /no-integration /ibc /no-clearnugetcache /configuration $(BuildConfiguration)
+  displayName: Build ProjectSystem.sln
+
+- task: RichCodeNavIndexer@0
+  displayName: RichCodeNav Upload
+  inputs:
+    languages: 'csharp'
```
Comment (dmonroym) [inline build/ci/richnav.yml:50]: Just to have a better understanding, we don't need to explicitly specify VB here?

**Comment by dmonroym** [comment]
@jepetty I also just re-read the description and think it would be great to be added to the preview of integrated GitHub C# navigation - how do we do this for the whole team?

File: build/ci/richnav.yml (hunk 4 above)
Comment (jepetty) [reply build/ci/richnav.yml:50]: Nope! When you pass csharp it indexes both csharp and VB

File: build/ci/unit-tests-template.yml (hunk 3 above)
Comment (jepetty) [reply build/ci/unit-tests-template.yml]: Done! Thanks

**Comment by jepetty** [comment]
@dmonroym followed up with me offline in onboarding the project team to the integrated c# navigation experience

**Review by dmonroym — APPROVED** [review]

**Pull request merged**